Add household model calculations

Eleonore9 opened the pull request:
- add testing data in `resources/testing_data.edn`
- add schemas in `src/witan/models/schemas.clj`
- add helpers in `src/witan/models/utils.clj`
- fill in the workflowfns in `src/witan/models/household.clj`
- add tests using the testing data in `test/witan/models/household_test.clj

Commit:
Create simple testing datasets

Commit:
Add testing data, tests and calculations

Add test for calc-hh-popn workflowfn

Add 2nd calc fn + test, testing data in edn file

Add data, calc, test for housholds calc.

Improve test to compare values w/ `fp-equals?`

Add data, calc and test for total households

Add data, calc, test for occupancy rates calc

Add data, calc, test for dwellings calc

Eleonore9 commented:
:( will look at my tests tomorrow! They seem to pass locally... To investigate

Eleonore9 commented:
Ok, that's normal. The `household-workspace-test` will fail until I add all the calculations.

Commit:
Fix the tests

Commit:
Format w/ lein cljfmt

acron0 commented on resources/testing_data.edn:
```diff
@@ -0,0 +1,70 @@
+{:test-inputs
```
Could this file be a separate resources dir which is referenced in the `dev` profile? Otherwise this will be included in the JAR.

acron0 commented on src/witan/models/household.clj:
```diff
@@ -47,7 +50,15 @@
    :witan/output-schema {:household-popn s/HouseholdPopulation}}
   [{:keys [resident-popn institutional-popn]} _]
-  {:household-popn {}})
+  (let [joined-popns (wds/join resident-popn institutional-popn
```
This `let` can be `thread-first`d (`->`)

acron0 commented on src/witan/models/household.clj:
```diff
@@ -57,7 +68,12 @@
    :witan/output-schema {:household-popn-grp s/HouseholdPopulationGrouped}}
   [{:keys [household-popn]} _]
-  {:household-popn-grp {}})
+  (let [popn-with-grp-col (wds/add-derived-column household-popn :age-group
```
Thread first again

acron0 commented on src/witan/models/household.clj:
```diff
@@ -57,7 +68,12 @@
-  {:household-popn-grp {}})
+  (let [popn-with-grp-col (wds/add-derived-column household-popn :age-group
+                                                  [:age]
+                                                  (fn [a] (u/get-age-grp a)))
```
Don't need the surrounding `fn` part, just use `u/get-age-grp`

acron0 commented on src/witan/models/household.clj:
```diff
@@ -67,8 +83,16 @@
-  [{:keys [household-rates household-popn-grp]} _]
-  {:households {}})
+  [{:keys [household-representative-rates household-popn-grp]} _]
+  (let [joined-ds (wds/join household-representative-rates household-popn-grp
```
Thread first again

acron0 commented on src/witan/models/household.clj:
```diff
@@ -78,29 +102,38 @@
+   :witan/output-schema {:occupancy-rates s/OccupancyRates}}
   [{:keys [vacancy-rates second-homes-rates]} _]
-  {:occupancy-rate {}})
+  (let [joined-ds (wds/join vacancy-rates second-homes-rates [:gss-code :year])
```
Thread first :)

acron0 commented on src/witan/models/household.clj:
```diff
@@ -78,29 +102,38 @@
+  (let [joined-ds (wds/join total-households occupancy-rates [:gss-code :year])
+        dwellings-calc (wds/add-derived-column joined-ds :dwellings
+                                               [:households :occupancy-rates]
+                                               (fn [hh or] (* hh (- 1 or))))
```
`or` is a bad name for a value because it conflicts with `clojure.core/or` -> you can suffix with `'` (e.g. `or'` if you really want to use that initialisation). Also, `(- 1 or)` (i.e. `1 - or`) is definitely what you want here?

Eleonore9 replied:
Oh thanks. I just went for `or` bc it stands for "occupancy rates".
Yes, what I went is `housholds * (1 - occupancy-rates)`

acron0 commented on src/witan/models/household.clj:
```diff
@@ -78,29 +102,38 @@
-  [{:keys [total-households occupancy-rate]} _]
-  {:dwellings {}})
+  [{:keys [total-households occupancy-rates]} _]
+  (let [joined-ds (wds/join total-households occupancy-rates [:gss-code :year])
```
Thread first

Commit:
Implement @acron0 's comments

Remove data out of the resources dir
Replace `let`s by `->`, chge variable name

sunnytownsend commented:
General question - can you remind me why do we use edn files?

sunnytownsend commented on data/testing_data.edn:
```diff
@@ -0,0 +1,70 @@
+
+  :input-second-homes-rates [{:gss-code "001" :year 2015 :second-homes-rates 0.153}]}
+
+ :test-outputs
```
where did you get these from?

Eleonore9 replied:
They're all fake data

Eleonore9 commented:
Edn files are easy to read in as Clojure data structure.
In this case, it's temporary. Until I get "real" data.

sunnytownsend commented:
we discussed that in the next branch, the fake data would be replaced with extracts of real data from Pier's model to test the functions we are implementing are doing what he intended. For this PR though - lgtm :shipit:
